algorithms/linfa-clustering/src/k_means/init.rs

```diff
@@ -0,0 +1,389 @@
+        };
+        assert!(
+            abs_diff_eq!(out, expected_centroids) || {
+                expected_centroids.invert_axis(Axis(0));
+                abs_diff_eq!(out, expected_centroids)
+            }
+        );
+    }
+
+    #[test]
+    fn test_k_means_plusplus() {
+        verify_init(KMeansInit::KMeansPlusPlus);
+    }
+
+    // Using Isaac here is quite slow but yields reproduceable results
+    fn isaac_rng(seed: u64) -> f64 {
+        let mut rng = Isaac64Rng::seed_from_u64(seed);
+        rng.gen_range(0.0, 1.0)
+    }
+
+    #[test]
+    fn test_k_means_para() {
+        verify_init(KMeansInit::KMeansPara(isaac_rng as RngFunc<f64>));
+    }
+
+    // Run general tests for a given init algorithm
+    fn verify_init(init: KMeansInit<f64>) {
+        let mut rng = Isaac64Rng::seed_from_u64(42);
+        // Make sure we don't panic on degenerate data (n_clusters > n_samples)
+        let degenerate_data = array![[1.0, 2.0]];
+        let out = init.run(2, &degenerate_data.view(), &mut rng);
+        assert_abs_diff_eq!(out, stack![Axis(0), degenerate_data, degenerate_data]);
+
+        // Build 3 separated clusters of points
+        let centroids = [20.0, -1000.0, 1000.0];
+        let clusters: Vec<Array2<_>> = centroids
+            .iter()
+            .map(|&c| Array::random_using((50, 2), Normal::new(c, 1.).unwrap(), &mut rng))
+            .collect();
+        let obs = clusters.iter().fold(Array2::default((0, 2)), |a, b| {
+            stack(Axis(0), &[a.view(), b.view()]).unwrap()
+        });
+
+        // Look for the right number of centroids
+        let out = init.run(centroids.len(), &obs.view(), &mut rng);
+        let mut cluster_ids = HashSet::new();
+        for row in out.genrows() {
+            // Centroid should not be 0
+            assert_abs_diff_ne!(row, Array1::zeros(row.len()), epsilon = 1e-1);
+            // Find the resultant centroid in 1 of the 3 clusters
+            let found = clusters
+                .iter()
+                .enumerate()
+                .filter_map(|(i, c)| {
+                    if c.genrows().into_iter().any(|cl| abs_diff_eq!(row, cl)) {
+                        Some(i)
+                    } else {
+                        None
+                    }
+                })
+                .next()
+                .unwrap();
+            cluster_ids.insert(found);
+        }
+        // Centroids should almost always span all 3 clusters
+        assert_eq!(cluster_ids, [0, 1, 2].iter().copied().collect());
+    }
+
+    #[test]
+    fn test_compare() {
+        let mut rng = Isaac64Rng::seed_from_u64(42);
+        let centroids = [20.0, -1000.0, 1000.0];
+        let clusters: Vec<Array2<_>> = centroids
+            .iter()
+            .map(|&c| Array::random_using((50, 2), Normal::new(c, 1.).unwrap(), &mut rng))
+            .collect();
+        let obs = clusters.iter().fold(Array2::default((0, 2)), |a, b| {
+            stack(Axis(0), &[a.view(), b.view()]).unwrap()
```

Comment:
this copies in every step, but the matrices are so small that it should not be a problem

Reply:
Yeah I'm not really concerned about perf in unit tests.